toolchain/base/value_store.h

```diff
@@ -223,6 +221,69 @@ auto CanonicalValueStore<IdT>::Reserve(size_t size) -> void {
   values_.Reserve(size);
 }
 
+// A ValueStore that builds a 1:1 relationship between two IDs.
+// * `RelatedIdT` represents a related ID that can be used to find values in the
+//   store.
+// * `IdT` is the actual ID of values in this store, and `IdT::ValueType` is the
+//   value type being stored.
+//
+// The value store builds a mapping so that either ID can be used later to find
+// a value. And the user can query if a related `RelatedIdT` has been used to
+// add a value to the store or not.
+//
+// When adding to the store, the user provides the related `RelatedIdT` along
+// with the value being stored, and gets back the ID of the value in the store.
+template <typename RelatedIdT, typename IdT>
+class RelationalValueStore {
+ public:
+  using ValueType = IdT::ValueType;
+  using RefType = ValueStore<IdT>::RefType;
+  using ConstRefType = ValueStore<IdT>::ConstRefType;
+
+  // Given the related ID and a value, stores the value and returns a mapped ID
+  // to reference it in the store.
+  auto Add(RelatedIdT related_id, ValueType value) -> IdT {
+    CARBON_DCHECK(related_id.index >= 0, "{0}", related_id);
+    IdT id(related_id.index);
+    if (static_cast<size_t>(id.index) >= values_.size()) {
+      values_.resize(id.index + 1);
+    }
+    values_[id.index] = std::move(value);
+    return id;
+  }
+
+  // Returns the ID of a value in the store if the `related_id` was previously
+  // used to add a value to the store, or None.
+  auto TryGetId(RelatedIdT related_id) const -> IdT {
+    CARBON_DCHECK(related_id.index >= 0, "{0}", related_id);
+    if (static_cast<size_t>(related_id.index) >= values_.size()) {
+      return IdT::None;
+    }
+    auto& opt = values_[related_id.index];
+    if (!opt.has_value()) {
+      return IdT::None;
+    }
+    return IdT(related_id.index);
+  }
+
+  // Returns a mutable value for an ID.
+  auto Get(IdT id) -> RefType {
```

Comment:
Is this function used?

```suggestion
  auto MutableGet(IdT id) -> RefType {
```

Reply:
No, I was just copying from ValueStore, but removed.